File: Revario-Go/Day14/day14.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
)

var input = `ngcjuoqr`



func main() {

	fmt.Printf("Index Part 1: %v, Part 2: %v", find64th(false), find64th(true))

}

func find64th(stretch bool) int {
	keyNum := []int{}
	hashes := []string{}
	for i := 0; true; i++ {
		if len(hashes) < i+1000 {
			if stretch {
				hashes = append(hashes, generateHashes(len(hashes), 1001, true)...)
			} else {
				hashes = append(hashes, generateHashes(len(hashes), 1001, false)...)
			}
		}
		char, found := hasStartSeq(hashes[i])
		if found {
			for _, vv := range hashes[i+1 : i+1001] {
				if hasEndSeq(char, vv) {

					keyNum = append(keyNum, i)
					break
				}
			}
		}

		if len(keyNum) == 64 {
			return keyNum[63]
		}
	}
	return -1
}
func hasStartSeq(hash string) (byte, bool) {
	for i := 0; i < len(hash)-2; i++ {
		if strings.Count(string(hash[i:i+3]), string(hash[i])) == 3 {
			return hash[i], true
		}
	}
	return '0', false
}

func hasEndSeq(keyChar byte, hash string) bool {
	for i := 0; i < len(hash)-5; i++ {
		if strings.Count(string(hash[i:i+5]), string(keyChar)) == 5 {
			return true
		}
	}
	return false
}

func generateHashes(start, amount int, stretched bool) []string {

	hashes := []string{}
	for i := 0; i < amount; i++ {

		hash := md5.Sum([]byte(input + strconv.Itoa(start+i)))
		hexa := hex.EncodeToString([]byte(hash[:]))
		hashes = append(hashes, hexa)

		if stretched {
			for a := 0; a < 2016; a++ {
				hash := md5.Sum([]byte(hashes[i]))
				hexa := hex.EncodeToString([]byte(hash[:]))
				hashes[i] = hexa
			}
		}
	}
	return hashes
}

File: dlofstrom-python/Day14/day14.py
from hashlib import md5
import re

input = 'jlmsuwbz'

i3 = 0
keys = []
while len(keys)<64:
    h3 = md5(input+str(i3)).hexdigest()
    m3 = re.findall(r'((\w)\2\2)',h3)
    if m3:
        for i5 in range(i3,i3+1000):
            h5 = md5(input+str(i5)).hexdigest()
            m5 = re.findall(r'((\w)\2\2\2\2)',h5)
            if m5:
                if m3[0][1] in [m[0][1] for m in m5] and not i3==i5:
                    keys += [i3]
                    break
    #print index, hash
    i3 += 1

print 'Part 1:', keys[-1]
